legislation_analysis/processing/ner.py

```diff
@@ -0,0 +1,238 @@
+
+    @classmethod
+    def aggregate_ner(cls, entities: list) -> dict:
+        """
+        Aggregates Named Entity Recognition (NER) data.
+        Conductions aggregation by embedding NER into sentence vectors,
+        computing the cosine similarity between the vectors, and aggregating
+        similar NERs using the longest entity name.
+
+        Parameters:
+            ner_lst (list): List of Named Entity Recognition (NER) data.
+
+        Returns:
+            ner_dict (dict): Aggregated Named Entity Recognition (NER) data.
+        """
+        # Group entities by their label
+        entities_by_label = cls.group_ent_by_label(entities)
+
+        # Generate Embeddings for Each Group
+        embeddings_by_label = {}
+        for label, names in entities_by_label.items():
+            embeddings_by_label[label] = model.encode(names)
+
+        # Compute Similarity and Aggregate Within Groups
+        threshold = 0.70
+        canonical_entities = defaultdict(int)
+        for label, embeddings in embeddings_by_label.items():
+            similarity_matrix = cosine_similarity(embeddings)
+
+            # Flag to mark entities that have been aggregated to avoid double
+            # counting
+            aggregated = [False] * len(embeddings)
+
+            for i in range(len(embeddings)):
+                # Skip already aggregated entities
+                if aggregated[i]:
+                    continue
+
+                aggregated[i] = True
+
+                # Initialize aggregation with the current entity
+                similar_entities = [entities_by_label[label][i]]
+
+                for j in range(i + 1, len(embeddings)):
+                    if similarity_matrix[i, j] > threshold:
+                        similar_entities.append(entities_by_label[label][j])
+                        aggregated[j] = True  # Mark as aggregated
+
+                # Choose the canonical name as the longest name among similar
+                # entities
+                canonical_name = max(similar_entities, key=len)
+                canonical_entities[(canonical_name, label)] += len(
+                    similar_entities
+                )
+
+        # convert to list of tuples
+        canonical_entities_tuples = []
+        for (name, label), count in canonical_entities.items():
+            canonical_entities_tuples.append((name, label, str(count)))
+
+        return canonical_entities_tuples
+
+    def process(self, cols_to_ner=None) -> None:
+        """
+        Process the text of the legislation to apply Named Entity Recognition
+        (NER).
+
+        parameters:
+            cols_to_ner (list): columns to apply NER to.
+        """
+        if cols_to_ner is None:
+            cols_to_ner = [("cleaned_text", "cleaned_text_ner")]
+        self.ner_df = self.df.copy()
+
+        # Apply NER to the text of the legislation
+        for col in cols_to_ner:
+            logging.debug(f"\tApplying NER to {col[0]}...")
```

Comment:
Thoughts on a single period version `...`?

Reply:
i think elipsis is nice